### ISSUE TITLE
[Feature Request] Keras ops.gradient function

### ISSUE THREAD

**Issue by PatReis** [keras-team/keras#19220]
Hello,
I was wondering whether it would be possible to provide a keras ops grad function based on e.g. the jax design.
This would be helpful to implement energy/force models or PiNNs in keras 3.
I remember that the first keras initially had a backend gradient function.
I already tried and it worked for tensorflow and torch but for Jax models the gradient output did not work in training for me.
Is this something you would be interested in? Do you see potential problems with jax?
Thanks!

**Comment by fchollet** [comment]
This is not straightforward in a multi-backend setting. We could make it work for Keras graphs (Functional models) but not for arbitrary backend-level computation.

Why not just use backend specific APIs like `tf.GradientTape` or `jax.grad`?

**Comment by PatReis** [comment]
Okay. Do you mean something linke this (see below). 
I just wanted to leave it in case somebody is looking for a similar issue.
I tested and it seems to work for all backends. 

If you do not see an obvious problem with keras API you can close this issue.

```python
import keras as ks
import numpy as np
from keras.backend import backend

model = ks.models.Sequential()
model.add(ks.layers.Dense(32, activation="swish"))
model.add(ks.layers.Dense(32, activation="swish"))
model.add(ks.layers.Dense(1, activation="linear"))


class FirstDeriv(ks.models.Model):

    def __init__(self, m):
        super(FirstDeriv, self).__init__()
        self.m = m

    def call(self, inputs, **kwargs):

        if backend() == "torch":
            import torch
            with torch.enable_grad():
                inputs.requires_grad = True
                out = self.m.call(inputs, **kwargs)
                out_sum = torch.sum(out, dim=0)[0]
                out_grad = torch.unsqueeze(torch.autograd.grad(out_sum, inputs, create_graph=True)[0], dim=-1)
        elif backend() == "tensorflow":
            import tensorflow as tf
            with tf.GradientTape(persistent=True) as tape:
                tape.watch(inputs)
                out = self.m.call(inputs, **kwargs)
                out_sum = tf.reduce_sum(out, axis=0, keepdims=False)[0]
            out_grad = tf.expand_dims(tape.gradient(out_sum, inputs), axis=-1)
        elif backend() == "jax":
            import jax
            import jax.numpy as jnp
            out = self.m.call(inputs, **kwargs)
            def fn_out(inputs):
                out_call = self.m.call(inputs, **kwargs)
                out_sum = jnp.sum(out_call, axis=0)[0]
                return out_sum
            grad_fn = jax.grad(fn_out)
            out_grad = grad_fn(inputs)

        else:
            raise NotImplementedError()

        return out, out_grad


X = np.random.normal(2, 3, size=(20,1))
y = [np.square(X), 2*X]

model_grad = FirstDeriv(model)
model_grad.predict(X)
model_grad.compile(optimizer=ks.optimizers.Adam(), loss=["mean_absolute_error", "mean_absolute_error"])
model_grad.fit(X, y, epochs=1000, verbose=2)
```

**Comment by fchollet** [comment]
Thanks for the code snippet. Closing as there are no plans to add a backend-agnostic gradient API at this time.
